docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +149,203 @@ Cluster Service Administration
 +----------+---------------+-------------+----------------------------+
 | edit     | dex-config    | name        | open an editor             |
 +----------+---------------+-------------+----------------------------+
+
+Design Choices
+--------------
+
+Two distinct CLI:
+
+- a ``metalk8s`` kubectl plugin with subcommands to interact with Kubernetes
+  API, and Salt API if needed.
+- a ``metalk8sctl`` CLI with subcommands for action that need to interact with
+  the local machine, but may also interact with Kubernetes API and Salt API if
+  needed.
+
+
+``metalk8s`` kubectl plugin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Language
+""""""""
+
+``Go`` is chosen as the language for kubectl plugin for the following reasons:
+
+- Great interactions with Kubernetes API.
+- Often used for operators and kubectl plugins
+  (`Sample CLI plugin <https://github.com/kubernetes/sample-cli-plugin>`_).
+- Portable as it's just a binary.
+- Simple deployment (no real requirements), just drop a binary in the PATH.
+
+Input and Output
+""""""""""""""""
+
+Each command should follow the kubectl style and standard as much as possible:
+
+- Command style:
+
+  .. code-block:: bash
+
+    kubectl metalk8s <action> <resource>
+
+- Interactive:
+
+  No interaction with the user, except when it's an ``edit`` command an editor
+  is opened (if needed).
+
+- Output style:
+
+  Default human-readable output (``<object> <action>ed``).
+
+  A ``--output``, ``-o`` option to change output format, at least support
+  for ``json`` and ``yaml`` (``jsonpath`` and ``go-template`` when it's
+  possible).
+
+  Internally each action result should be an "object"
+  (e.g.: single-level dictionary) containing several informations,
+  at least:
+
+  - name
+  - message
+  - result (True or False)
+  - an elapsed time (to know each action time)
+
+  A ``--verbose``, ``-v`` option to change log level verbosity (default output
+  to ``stderr``), using
+  `Kubernetes log library <https://github.com/kubernetes/klog>`_.
+
+
+SaltAPI interaction
+"""""""""""""""""""
+
+If the plugin needs to access Salt API a proxy should be configured by the
+plugin.
+
+And then plugin can access Salt API directly using the proxy
```

Comment:
For what I understand we do not want to hit directly apiserver to reach this proxy as you need to be authenticated through apiserver to hit this proxy but you also need to authenticate on salt-api so .... For me we want to proxy apiserver

Reply:
> we want to proxy apiserver

But we're already authenticated to apiserver, and can already talk to it, since we're in a `kubectl` plugin, right?
And we can pass the token we already have from apiserver to Salt API, that should be enough.
I don't understand what we are missing...

Reply:
Ok just did some test it's doable giving two token on every call, right I will talk about it here

```
curl -kvvi -X POST https://10.100.2.176:6443/api/v1/namespaces/kube-system/services/https:salt-mster:api/proxy/ -H "Authorization: Bearer $id_token" -H "X-Auth-Token: $salt_token" -d 'client=local' -d 'tgt=*' -d 'fun=test.ping'
```

Reply:
Done